# 01. Importing Libraries

In [6]:
#Import the pandas, NumPy, and os libraries.
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [8]:
#Creating a path
path = r'C:\Users\canam\Achievement 4\Instacart Basket Analysis'

In [10]:
#Importing product.csv 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [24]:
#Importing orders_wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Mixed Type Data

In [14]:
#Create dataframe
df_test = pd.DataFrame()

In [15]:
#Create mixed type column 
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [17]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [18]:
#converting data type into string 
df_test['mix'] = df_test['mix'].astype('str')

# 04. Missing values 

In [20]:
#Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [21]:
#Creating a subset of the missing values 
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [22]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [23]:
df_prods.shape

(49693, 5)

In [24]:
#creating a subset of non-missing values 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

Another way to drop missing values is df_prods.dropna(inplace = True).  To to drop only the NaNs from a particular column, the code would look like this:

df_prods.dropna(subset = [‘product_name’], inplace = T’re

# 05. Duplicates 

In [27]:
#looking for duplicates 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [28]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [29]:
df_prods_clean.shape

(49677, 5)

In [30]:
#getting rid of duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [31]:
df_prods_clean_no_dups.shape

(49672, 5)

# 06. Exporting the new products dataframe

In [33]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 07. Task: Cleaning up df_ords dataframe

In [26]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

The column 'days_since_last_order' has missing values.  Its count is lower than that  of the other columns.  Some orders were placed on the same day as the minimum 'days_since_last_order' is 0.  Order_day_of_week seems to be in order, with a min of 0 and a max of 6 for the 7 days of the week.  Same thing with the 'order_hour_of_day' which has a min of 0 and a maximum of 23 for the 24 hours in a day.

In [16]:
#Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are no columns with mixed-type data in the df_ords dataframe.

In [43]:
#Run a check for missing values in your df_ords dataframe.  Report your findings and propose an explanation for any missing values you find.
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

There are 206209 values missing in the column days_since_last_order.  This either perhaps because those values were never recorded or there was a technical glitch.  We need to view these missing values.  

In [45]:
#creating a subset of missing values
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull()==True]

In [47]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [12]:
#checking to see if all order_number_1s are missing values for days_since_last_order
df_order_number_1 =  df_ords[df_ords['order_number']==1]

In [14]:
df_order_number_1

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

The number of 'order_numbers_1's is the same as the number of missing values, so we can assume that all order_number_1s have a missing value for days_since_last_order.  This makes sense because there can't be any days_since_last_order for the first order.  Instead of deleting the rows with the missing values (which would remove more than 5% of the data), we will create a new variable (first_order) that acts like a flag based on the missing value.

In [20]:
#creating a new variable that acts like a flag based on the missing value 
df_ords['first_order'] = df_ords.isna().any(axis=1).astype(int)

In [18]:
#checking to see that the new variable was created
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_last_order  first_order  
0                  8                    NaN            1  
1                  7                   15.0            0  
2                 12                   21.0            0  
3                  7                   29.0            0  
4                 15                   28.0            0

In [21]:
#Run a check for duplicate values in your df_ords data.  
#In a markdown cell, report your findings and propose an explanation for any duplicate values you find.'

df_dups = df_ords[df_ords.duplicated()]

In [23]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

There are no duplicates in the df_ords dataframe.  

In [22]:
#exporting final cleaned df_ords as .csv.  
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))